src/Compilers/CSharp/Portable/Symbols/Source/ExplicitInterfaceHelpers.cs

```diff
@@ -17,21 +17,78 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
+#nullable enable
+    internal class ExplicitInterfaceMemberInfo(Binder binder, BindingDiagnosticBag diagnostics, ExplicitInterfaceSpecifierSyntax explicitInterfaceSpecifier, string name, string? aliasQualifier)
+    {
+        private TypeSymbol? _lazyExplicitInterfaceType;
+        private string? _lazyMemberMetadataName;
+
+        public ExplicitInterfaceSpecifierSyntax ExplicitInterfaceSpecifier => explicitInterfaceSpecifier;
+
+        public TypeSymbol? ExplicitInterfaceType
+        {
+            get
+            {
+                if (_lazyExplicitInterfaceType is null)
+                {
+                    // Avoid checking constraints context when binding explicit interface type since
+                    // that might result in a recursive attempt to bind the containing class.
+                    var typeBinder = binder.WithAdditionalFlags(BinderFlags.SuppressConstraintChecks | BinderFlags.SuppressObsoleteChecks);
+
+                    NameSyntax explicitInterfaceName = explicitInterfaceSpecifier.Name;
+                    TypeSymbol? explicitInterfaceType = binder.BindType(explicitInterfaceName, diagnostics).Type;
+                    return InterlockedOperations.Initialize(ref _lazyExplicitInterfaceType, explicitInterfaceType);
+                }
+
+                return _lazyExplicitInterfaceType;
+            }
+        }
+
+        public string MemberMetadataName
+        {
+            get
+            {
+                if (_lazyMemberMetadataName is null)
+                {
+                    var memberMetadataName = ExplicitInterfaceHelpers.GetMemberMetadataName(name, ExplicitInterfaceType, aliasQualifier);
+                    return InterlockedOperations.Initialize(ref _lazyMemberMetadataName, memberMetadataName);
+                }
+
+                return _lazyMemberMetadataName;
+            }
+        }
+    }
+
     internal static class ExplicitInterfaceHelpers
     {
         public static string GetMemberName(
-            Binder binder,
-            ExplicitInterfaceSpecifierSyntax explicitInterfaceSpecifierOpt,
+            ExplicitInterfaceSpecifierSyntax? explicitInterfaceSpecifier,
             string name)
         {
-            TypeSymbol discardedExplicitInterfaceType;
-            string discardedAliasOpt;
-            string methodName = GetMemberNameAndInterfaceSymbol(binder, explicitInterfaceSpecifierOpt, name, BindingDiagnosticBag.Discarded, out discardedExplicitInterfaceType, out discardedAliasOpt);
+            string? aliasQualifier = explicitInterfaceSpecifier?.Name.GetAliasQualifierOpt();
+            return GetMemberName(name, explicitInterfaceSpecifier?.Name.ToString(), aliasQualifier);
```

Comment:
>explicitInterfaceSpecifier?.Name.ToString() [](http://example.com/codeflow?start=39&length=43)

I think we should make sure this doesn't include any trivia like comments, pragmas, pound directives, etc.